File: frontend/lib/blackScholes.ts
function normCdf(x: number): number {
  const a = 0.3275911;
  const b = [0.254829592, -0.284496736, 1.421413741, -1.453152027, 1.061405429];
  const sign = x < 0 ? -1 : 1;
  const t = 1 / (1 + a * Math.abs(x));
  const poly = b.reduce((acc, bi) => acc * t + bi, 0) * t;
  return 0.5 * (1 + sign * (1 - poly * Math.exp(-x * x / 2)));
}

function normPdf(x: number): number {
  return Math.exp(-x * x / 2) / Math.sqrt(2 * Math.PI);
}

function d1(S: number, K: number, T: number, r: number, sigma: number) {
  return (Math.log(S / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * Math.sqrt(T));
}

function d2(d1val: number, sigma: number, T: number) {
  return d1val - sigma * Math.sqrt(T);
}

export type OptionResult = {
  price: number;
  delta: number;
  gamma: number;
  vega: number;
  theta: number;
  rho: number;
};

export function price(
  S: number, K: number, T: number, r: number, sigma: number,
  type: "call" | "put"
): OptionResult {
  const _d1 = d1(S, K, T, r, sigma);
  const _d2 = d2(_d1, sigma, T);
  const disc = Math.exp(-r * T);
  const sqrtT = Math.sqrt(T);

  const callPrice = S * normCdf(_d1) - K * disc * normCdf(_d2);

  return {
    price:  type === "call" ? callPrice : callPrice - S + K * disc,
    delta:  type === "call" ? normCdf(_d1) : normCdf(_d1) - 1,
    gamma:  normPdf(_d1) / (S * sigma * sqrtT),
    vega:   S * sqrtT * normPdf(_d1) / 100,
    theta:  ((-S * normPdf(_d1) * sigma / (2 * sqrtT))
              + (type === "call" ? -1 : 1) * r * K * disc * normCdf(type === "call" ? _d2 : -_d2)) / 365,
    rho:    (type === "call" ? 1 : -1) * K * T * disc * normCdf(type === "call" ? _d2 : -_d2) / 100,
  };
}

File: src/black_scholes.py
"""Black-Scholes pricer — European options."""

import math
from scipy.stats import norm


def _d1(S: float, K: float, T: float, r: float, sigma: float) -> float:
    """d1 : log-moneyness ajusté de la dérive et de la convexité."""
    return (math.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * math.sqrt(T))


def _d2(d1: float, sigma: float, T: float) -> float:
    """d2 = d1 - sigma*sqrt(T) : probabilité risque-neutre d'exercice."""
    return d1 - sigma * math.sqrt(T)


def bs_price(
    S: float, K: float, T: float, r: float, sigma: float,
    option_type: str = "call",
) -> float:
    """
    Prix Black-Scholes d'une option européenne.

    Parameters
    ----------
    S           : prix spot du sous-jacent
    K           : prix d'exercice (strike)
    T           : maturité en années
    r           : taux sans risque continu
    sigma       : volatilité annualisée
    option_type : "call" ou "put"
    """
    d1 = _d1(S, K, T, r, sigma)
    d2 = _d2(d1, sigma, T)
    discount = math.exp(-r * T)

    if option_type == "call":
        return S * norm.cdf(d1) - K * discount * norm.cdf(d2)
    elif option_type == "put":
        # Parité call-put : P = C - S + K*e^(-rT)
        call = S * norm.cdf(d1) - K * discount * norm.cdf(d2)
        return call - S + K * discount
    else:
        raise ValueError(f"option_type doit être 'call' ou 'put', reçu '{option_type}'")


def bs_delta(S: float, K: float, T: float, r: float, sigma: float,
             option_type: str = "call") -> float:
    """Delta = dC/dS.  Call : N(d1).  Put : N(d1) - 1."""
    d1 = _d1(S, K, T, r, sigma)
    if option_type == "call":
        return norm.cdf(d1)
    return norm.cdf(d1) - 1.0


def bs_gamma(S: float, K: float, T: float, r: float, sigma: float) -> float:
    """Gamma = d²C/dS² = N'(d1) / (S·sigma·sqrt(T)).  Identique call et put."""
    d1 = _d1(S, K, T, r, sigma)
    return norm.pdf(d1) / (S * sigma * math.sqrt(T))


def bs_vega(S: float, K: float, T: float, r: float, sigma: float) -> float:
    """Vega = dC/d_sigma = S·sqrt(T)·N'(d1).  Pour un move de 1% de vol."""
    d1 = _d1(S, K, T, r, sigma)
    return S * math.sqrt(T) * norm.pdf(d1) / 100


def bs_theta(S: float, K: float, T: float, r: float, sigma: float,
             option_type: str = "call") -> float:
    """Theta = dC/dt par jour calendaire (convention 365 jours)."""
    d1 = _d1(S, K, T, r, sigma)
    d2 = _d2(d1, sigma, T)
    decay = -S * norm.pdf(d1) * sigma / (2 * math.sqrt(T))
    discount = math.exp(-r * T)

    if option_type == "call":
        return (decay - r * K * discount * norm.cdf(d2)) / 365
    return (decay + r * K * discount * norm.cdf(-d2)) / 365


def bs_rho(S: float, K: float, T: float, r: float, sigma: float,
           option_type: str = "call") -> float:
    """Rho = dC/dr.  Pour un move de 1% du taux."""
    d1 = _d1(S, K, T, r, sigma)
    d2 = _d2(d1, sigma, T)
    discount = math.exp(-r * T)

    if option_type == "call":
        return K * T * discount * norm.cdf(d2) / 100
    return -K * T * discount * norm.cdf(-d2) / 100
